tests/build system: allow buildtests to be executed according to MCU groups

phiros opened the pull request:
The idea behind this PR is to group buildtests by MCU type, thus allowing us to take advantage of the buildmatrix feature of travis (might also be useful if you want to 
execute a buildtest locally just for a certain group of boards).

An example test run can be seen here:
https://travis-ci.org/phiros/RIOT/builds/49693219
EDIT: also the test run on this PR ;-)

The whole thing is still very hacky but might be a reasonable short term solution until we have our own test cluster.
Also I feel like travis is generally very unreliably when it comes to jobs which take more than 10 min...

authmillenon commented on .travis.yml:
```diff
@@ -3,7 +3,11 @@ language: c
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=cortex_m3
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=cortex_m4
 
 before_install:
```
```diff
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=x86
```
?

gebart replied:
and avr?

gebart commented:
This is a good idea. Is it possible to reduce the number of packages to install on Travis initialization as well? Only install the toolchain for the targeted platform, I mean.

LudwigOrtmann commented:
@gebart good idea about optimizing installations, but maybe a good candidate for a follow up PR once this one is finished.

OlegHahm commented:
:+1: for this PR, for @gebart's proposal and for making it a follow-up PR.

phiros commented:
I have incorporated most of the feedback I got:
* The packages are now selectively installed according the MCU group (saves a lot of time)
* Unit tests are run only in the x86 group for now (unit tests currently run on x86 anyways)

In addition to that I split up a larger group of MCUs into two (cortex_m3 -> cortex_m3 and cortex_m3_2) in order to keep the run time of each build task low.

LudwigOrtmann commented on boards/redbee-econotag/Makefile.features:
```diff
@@ -1 +1,3 @@
 FEATURES_PROVIDED += transceiver
+FEATURES_MCU_GROUP = arm7
+
```
stray blank line

LudwigOrtmann commented on .travis.yml:
```diff
@@ -3,7 +3,14 @@ language: c
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=arm7
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=cortex_m0
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=cortex_m3
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=cortex_m3_2
```
please cortex_m3 => cortex_m3_1 for consistency

LudwigOrtmann commented:
Please rebase and check if any new boards popped up in master in the meantime.

phiros commented:
rebased (no new boards were added in the meantime) + fixed those two issues you mentioned.

gebart commented:
Is the cortex m3 split arbitrary or did you have any particular thoughts for which board went in which group? (just curious)

LudwigOrtmann commented on dist/tools/travis-scripts/build_and_test.sh:
```diff
@@ -0,0 +1,33 @@
+		#   invocation) once all warnings of cppcheck have been taken care of
+		#   in master.
+		./dist/tools/cppcheck/check.sh master --diff-filter=MR --error-exitcode=0 || exit
+		./dist/tools/cppcheck/check.sh master --diff-filter=AC || exit
```
The static analysis does not really depend on "$BUILDTEST_MCU_GROUP" == "x86", maybe you could factor it into either it's own group, or have it executed in a different fashion?

LudwigOrtmann commented on dist/tools/travis-scripts/build_and_test.sh:
```diff
@@ -0,0 +1,33 @@
+	if [ "$BUILDTEST_MCU_GROUP" == "x86" ]
+	then
+		make -s -C ./examples/default info-concurrency
+		git rebase riot/master || git rebase --abort
```
This is not specific to `x86`, should be done for all builds.

LudwigOrtmann commented:
@phiros ping

LudwigOrtmann commented:
Please rebase on master, there's a new script (from https://github.com/RIOT-OS/RIOT/pull/2468).

phiros commented:
@gebart the split for the cortex m3 boards is arbitrary.

@LudwigOrtmann The static checks execute now for every MCU group (IMHO the overhead is tolerable; each MCU group build takes ~1 min. longer to finish). 
Do I get an ACK from you for this? I think we can improve on this further with follow up PRs if we really
see the need for it.

LudwigOrtmann commented:
please squash

LudwigOrtmann commented:
(ACK for the content)

LudwigOrtmann commented:
I guess you want to remove the WIP label? You sound eager to merge this ;)

LudwigOrtmann commented:
Don't forget to remove the 'needs squashing' label once you squashed.

OlegHahm commented:
:raised_hands:
